include/rocksdb/utilities/object_registry.h

```diff
@@ -38,6 +39,23 @@ using FactoryFunc =
 // library
 using RegistrarFunc = std::function<int(ObjectLibrary&, const std::string&)>;
 
+// The signature of the function for loading plugins
+// Plugin contains the struct of properties associated with this Plugin.
+// Plugin is passed to the PluginFunc and the fields should be initalized
+// by the Plugin code.
+// On success, this function should return 0.  On failure, this function should
+// return non-zero and set the std::string* to an appropriate error message.
+using PluginFunc = std::function<int(Plugin*, std::string*)>;
+
+struct Plugin {
```

Comment:
What problem does this solve that is not solved by `std::function<int(ObjectLibrary&)>`? (Noting that passing non-const ref violates Google C++ guidelines)

Reply:
> 

There are a few reasons for the Plugin (as opposed to a simple RegistrarFunc):
1.  If we ever intend to do error checking or provide more information about a plugin, we need some place to hold it.  The Plugin struct was intended to be that place
2. The Registrar Func takes an argument.  I am not sure where I am supposed to get that from (or just pass in "")
3. Without this struct, how could one determine what was loaded or available to the application?  Using the Plugins and ObjectRegistry, we could add some useful information to tools to say what they are compiled with (e.g. this version of db_bench supports HDFS env).  Without the Plugin struct/name, We would not have access to this information.

Reply:
Also, this does not violate the Google C++ guidelines: "while non-optional output and input/output parameters should usually be references (which cannot be null)"